evaporate.js

```diff
@@ -432,10 +458,54 @@ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
            authorizedSend(complete);
         }
 
+        var numProcessed = 0,
+            numParts = -1;
+
+        function computePartMd5Digest(part) {
+           return function () {
+              var s = me.file.status;
+              if (s == ERROR || s == CANCELED) {
+                 return;
+              }
+
+              var md5_digest = con.cryptoMd5Method.call(this, this.result);
+
+              l.d(['part #', part.part, ' MD5 digest is ', md5_digest].join(''));
+              part.md5_digest = md5_digest;
+
+              delete part.reader; // release potentially large memory allocation
+
+              numProcessed += 1;
+
+              processPartsList();
+
+              if (numProcessed === numParts) {
+                 l.d('All parts have MD5 digests');
+              }
+
+              setTimeout(processPartsListWithMd5Digests, 1500);
```

Comment:
@tomsaffell this implementation isn't perfect in that it keeps iterating over the parts list each time BUT it is close to JIT because:

1. we start with the first part in `processPartsListWithMd5Digests()`. We only do one part.
2. when the data read is done, we calculate the md5 and hopefully allow GC to clean up memory
3. we attach the digest to the part and then kick off `processPartsList()` which will find only one part that is ready and will initiate upload of that part.
4. on return from `processPartsList()` we then wait 1500 seconds before we start to find the next part that needs an MD5 digest.
5. so while we're working the digest for the second part, we've got only one async PUT in progress. This happens until the browser says it has enough concurrent async jobs running (more than the option setting in EvaporateJS).
6. at a certain point, the browser is going to stop processing any new jobs until it has room to do so. At that point, EvaporateJS just has an MD5 calc job waiting...once that job runs, we then submit the PUT...and everything works cooperatively.

One of my previous implementations did not control the flow this way and I ended up with an async XHR queue full of CPU and IO intensive jobs to calculate the checksums. The PUT requests never had a chance. Interestingly, only Safari showed the XHR async jobs related to the contention were not visible in Chrome, only Safari. Wasn't I lucky?

So to summarize, the rules when computing checksums are:

1. Do not upload a part if its checksum isn't ready.
2. As soon as a part has a checksum, submit it for upload. Assuming the XHR queue is idle, upload should begin.
3. After upload is submitted (not necessarily executing), wait 1.5 seconds before trying to calculate the next part's checksum. This adds a job to the XHR queue. 
4. Until that XHR data request completes, nothing else happens except idle looping through the parts queue (this could be optimized, of course).
5. This XHR requests build up until the browser is no longer able to keep up, at which time the "next" XHR request to be processed once it has room, MUST be an XHR request for a data read.

There may be some missing edge cases here but the more I think about it, the more I think it's a solid approach. What's left is an optimization that doesn't keep iterative over a list of parts but works with, say, a FIFO queue of some kind, pushing/popping as needed.